man/meteodbdump.1
.TH METEODBDUMP "1" "October 2003" "Meteo station tools" Othello
.SH NAME
meteodbdump \- dump contents of meteo database
.SH SYNOPSIS
meteodbdump [
.BI \-b basename
] [
.BI \-c chunksize
] [
.B \-d
] [
.B \-r
] [
.BI \-f conffile
] [
.BI \-l logurl
]
.BI \-s station
.I table ...
.SH DESCRIPTION
The 
.BR meteodbdump (1)
programm dumps the contents of the meteo database (see
.BR meteo.sql (5))
to text files.
In contrast to ordinary
.BR mysqldump (1),
it usually reads the records of
a day at a time, i.e. less than 2000 rows,
so that the table does not stay locked for extended
time intervals.
Also the meteo tables tend to be so large that
.BR mysqldump (1)
will run out of memory before anything is dumped.
Furthermore, a dump into SQL statements is just too space consuming,
so there is an option to request a more compact load file format.

.BR meteodbdump (1)
connects to the database using the read only connection specified in
the either the default configuration file or the one specified with
the 
.B -f
option.
It then starts reading data for the station
.I station
and writes SQL insert statements to the files
.IB basename sdata
and
.IB basename avg.
Data is read in chunks of time length at most
.I chunksize
seconds. 

The SQL insert statements do not rely on a certain order of the fields
in the table definition. However, the contents of the
.B station
table and the
.B mfield
table are not dumped. Since stations and fields are completely described
only with this information, tools like
.BR mysql_dump (1)
should be used to backup those tables.

Without any additional arguments,
.BR meteodbdump (1)
dumps all data tables, i.e. avg and sdata.
The names of the tables to be dumped can be specified as additional
arguments, in which case only the specified tables are dumped.

.SH OPTIONS
.TP
.BI \-b,\-\-base= basename
The
.I basename
string is prepended to the strings
.I sdata 
and
.IR avg ,
which are then used as filenames for the output files. The 
.I basename
may contain slashes (/), but 
.BR meteodbdump (1)
will not create directories, so they should be created manually
before
.BR meteodump (1)
is run. Example: a
.I basename
of 
.I /var/tmp/Lajeado-
will let 
.BR meteodbdump (1)
dump the data to the files
.I /var/tmp/Lajeado-sdata
and
.IR /var/tmp/Lajeado-avg .

.TP
.BI \-c,\-\-chunksize= chunksize
read at most records for a time interval of length
.I chunksize 
seconds from the database at one time. By default, a days worth of records
is read in every iteration, minimizing
.BR metedbdump (1)'s
impact on concurrency. Larger values for the chunksize cause the dump to
use more memory, and usually hurt concurrency noticably.
.TP
.B \-d,\-\-debug
increase the debug level by one. Debug messages are written to standard
error.

.TP
.B \-r,\-\-raw
request raw format, i.e. tabulater separated fields, one line per row.
This reduces the output size by about a factor 3.

.TP
.BI \-f,\-\-config= conffile
use
.I conffile
as the meteo configuration file. See 
.BR meteo.xml (5)
for the configuration file syntax.

.TP
.BI \-l,\-\-logurl= logurl
send log messages to the destination described by the
.IR logurl . See
.BR meteo (1)
for a description of the possible formats for log URLs.

.TP
.BI \-s,\-\-station= station
name of the station to dump. This option can be given several times,
.BR meteodbdump (1)
will dump the data for each of the stations in turn. Data for all stations
will be written to the same files.

.SH BUGS
.BR meteodbdump (1)
terminates if a single query fails. This makes it impossible to dump
corrupted databases.

.SH "SEE ALSO"
.BR meteo (1),
.BR meteo.sql (5),
.BR meteo.xml (5)

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).



src/meteodbdump.cc
/*
 * meteodbdump.cc -- dump the meteo database for a given set of stations
 *                   to suitable load files, and provide load stations
 *                   that will load the files again
 *
 *                   This program is written so that it can be compiled and
 *                   run independently of the meteo library. I use it to
 *                   synchronize the database on my notebook with timon's.
 *                   It isn't documented in a manual page either, because it
 *                   is hardly useful anybody but me.
 *
 * (c) 2003 Dr. Andreas Mueller, Beratung und Entwicklung
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ChunkDumper.h>
#include <Configuration.h>
#include <MeteoException.h>
#include <StationInfo.h>
#include <string>
#include <vector>
#include <errno.h>
#include <strings.h>
#include <unistd.h>
#include <mdebug.h>
#include <fstream>
#include <getopt.h>

namespace meteo {
namespace dbdump {

static void	usage(const char *progname) {
	printf("usage: %s [ options ] tables \n", progname);
	printf("dump meteo database contents to a file. valid table names are avg and sdata\n");
	printf("options:\n");
	printf("  -f,--config=<conffile>  read <conffile> on startup instead of the default\n");
	printf("  -l,--logurl=<logurl>    send log messages to this url instead of stderr\n");
	printf("  -d,--debug              increase debug level\n");
	printf("  -s,--station=<station>  include data for this station\n"
	       "                          (multiple stations allowed\n");
	printf("  -b,--base=<base>        base path name, file namess will be built by\n");
	printf("  -c,--chunksize=<size>   dump data in chunks of size <size> seconds\n");
	printf("  -h,-?,--help            display this help message and exit\n");
	printf("  -r,--raw                write data raw instead of SQL insert statements\n");
}

static struct option	options[] = {
{ "help",	no_argument,		NULL,		'h' },
{ "config",	required_argument,	NULL,		'c' },
{ "logurl",	required_argument,	NULL,		'l' },
{ "debug",	no_argument,		NULL,		'd' },
{ "station",	required_argument,	NULL,		's' },
{ "base",	required_argument,	NULL,		'b' },
{ "chunksize",	required_argument,	NULL,		'c' },
{ "raw",	required_argument,	NULL,		'r' },
{ NULL,		0,			NULL,		 0  }
};

static void	main(int argc, char *argv[]) {
	stringlist	stations;
	std::string	basename("./");
	std::string	logurl("file:///-");
	std::string	conffile(METEOCONFFILE);
	int	chunksize = 86400; 	// one day worth of data
	int	c;
	bool	raw = false;

	// parse the command line
	int	longindex;
	while (EOF != (c = getopt_long(argc, argv, "db:l:s:f:c:rh?",
		options, &longindex)))
		switch (c) {
		case '?':
		case 'h':
			usage(argv[0]);
			return;
			break;
		case 'f':
			conffile = std::string(optarg);
			break;
		case 'l':
			logurl = std::string(optarg);
			break;
		case 'd':
			debug++;
			break;
		case 's':
			stations.push_back(std::string(optarg));
			break;
		case 'b':
			basename = std::string(optarg);
			break;
		case 'c':
			chunksize = std::stoi(optarg);
			break;
		case 'r':
			raw = !raw;
			break;
		}

	// set up logging
	mdebug_setup("meteopoll", logurl.c_str());
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "set up logging to %s",
		logurl.c_str());

	// open configuration
	Configuration	conf(conffile);

	// create a chunkdumper for basename and chunksize
	ChunkDumper	cd(basename, chunksize);

	// set some parameters for this station
	if (raw)
		cd.enableRaw();

	// if there are more arguments, use them to select the tables we
	// want to dump
	if (argc > optind) {
		cd.disableAvg(); cd.disableSdata();
		for (int i = optind; i < argc; i++) {
			if (0 == strcmp(argv[i], "avg"))
				cd.enableAvg();
			if (0 == strcmp(argv[i], "sdata"))
				cd.enableSdata();
		}
	}

	// do the work in a function, so that destructors are properly
	// called when it completes
	cd.dumpStations(stations);
}

} // namespace dbdump
} // namespace meteo

int	main(int argc, char *argv[]) {
	try {
		meteo::dbdump::main(argc, argv);
	} catch (meteo::MeteoException& me) {
		fprintf(stderr, "MeteoException in meteodbdump: %s/%s\n",
			me.getReason().c_str(), me.getAddinfo().c_str());
		exit(EXIT_FAILURE);
	}
	exit(EXIT_SUCCESS);
}
